### PR TITLE
Allow merging consecutive 'long' values in SortedRangeSet

### PR DESCRIPTION
**Pull request by rzeyde-varada**
This can be useful when pushing down TupleDomains generated by dynamic filtering.

Fixes https://github.com/trinodb/trino/issues/6076

### CODE REVIEW

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java:192]: This should be covered by `isAdjacent` instead of adding a separate method. The intent of `isAdjacent` is to indicate whether two ranges are touching each other. When the domain is integer numbers, two boundaries are adjacent if there can be no value in between.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java:427]: Why do we need this? It should be handled naturally by the call to `span` below.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java
Comment (rzeyde-varada) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java]: Good point, thanks!
Fixed in df37dfec30.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java
Comment (rzeyde-varada) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java]: You're right, we don't need it for correctness.
But after this change, some tests result in having an empty range in them, e.g.:
https://github.com/prestosql/presto/blob/df37dfec300192b6f33116a54d77ed89f91013c3/presto-spi/src/test/java/io/prestosql/spi/predicate/TestSortedRangeSet.java#L62-L71
Without skipping empty ranges, the resulting `rangeSet` will be equal to `[(1, 2)]` - which is correct, but IMHO can be optimized into `SortedRangeSet.none()`.
WDYT?

Comment (rzeyde-varada) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java]: P.S.
We can split "empty ranges' handling" into a separate PR, if preferable.

Comment (martint) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java]: Got it. Let's split it up into a separate commit within this PR to keep the logical changes separate.

Comment (rzeyde-varada) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/SortedRangeSet.java]: Sounds good - I have split the PR in to two commits:
```
1f57c13d00  Allow merging consecutive 'long' values in SortedRangeSet
c429fdd5d2  Skip empty ranges while constructing SortedRangeSet
```

**Comment by rzeyde-varada** [comment]
Two of the `TestTpcdsCostBasedPlan` tests failed, due to a join order flip:
[log link](https://gist.githubusercontent.com/rzeyde-varada/0ab0d7287aaa9f237d2e2f399379550c/raw/f10bd9bd3c3bddfe65c046a4b8c8bee6a9bdf24f/errors.txt).

**Comment by rzeyde-varada** [comment]
It also seems that statistics estimation is different after rewriting the query:
```
presto:tiny> EXPLAIN ANALYZE SELECT d_date_sk FROM date_dim WHERE d_year BETWEEN 1999 AND 2001;
                                                                                               Query Plan                                                                                               
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [SOURCE]                                                                                                                                                                                    
     CPU: 497.52ms, Scheduled: 527.01ms, Input: 73049 rows (0B); per task: avg.: 73049.00 std.dev.: 0.00, Output: 1096 rows (9.63kB)                                                                    
     Output layout: [d_date_sk]                                                                                                                                                                         
     Output partitioning: SINGLE []                                                                                                                                                                     
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                                                                      
     ScanFilterProject[table = tpcds:date_dim:sf0.01, grouped = false, filterPredicate = ("d_year" BETWEEN 1999 AND 2001)]                                                                              
         Layout: [d_date_sk:bigint]                                                                                                                                                                     
         Estimates: {rows: 73049 (642.03kB), cpu: 998.72k, memory: 0B, network: 0B}/{rows: 730 (6.42kB), cpu: 1.95M, memory: 0B, network: 0B}/{rows: 730 (6.42kB), cpu: 1.96M, memory: 0B, network: 0B} 
         CPU: 497.00ms (100.00%), Scheduled: 526.00ms (100.00%), Output: 1096 rows (9.63kB)                                                                                                             
         Input avg.: 18262.25 rows, Input std.dev.: 173.21%                                                                                                                                             
         d_date_sk := tpcds:d_date_sk                                                                                                                                                                   
         d_year := tpcds:d_year                                                                                                                                                                         
         Input: 73049 rows (0B), Filtered: 98.50%                                                                                                                                                       
                                                                                                                                                                                                        
                                                                                                                                                                                                        
(1 row)


presto:tiny> EXPLAIN ANALYZE SELECT d_date_sk FROM date_dim WHERE d_year IN (1999, 2000, 2001);
                                                                                                Query Plan                                                                                                
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [SOURCE]                                                                                                                                                                                      
     CPU: 591.47ms, Scheduled: 682.71ms, Input: 73049 rows (0B); per task: avg.: 73049.00 std.dev.: 0.00, Output: 1096 rows (9.63kB)                                                                      
     Output layout: [d_date_sk]                                                                                                                                                                           
     Output partitioning: SINGLE []                                                                                                                                                                       
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                                                                        
     ScanFilterProject[table = tpcds:date_dim:sf0.01, grouped = false, filterPredicate = ("d_year" IN (1999, 2000, 2001))]                                                                                
         Layout: [d_date_sk:bigint]                                                                                                                                                                       
         Estimates: {rows: 73049 (642.03kB), cpu: 998.72k, memory: 0B, network: 0B}/{rows: 1090 (9.58kB), cpu: 1.95M, memory: 0B, network: 0B}/{rows: 1090 (9.58kB), cpu: 1.96M, memory: 0B, network: 0B} 
         CPU: 591.00ms (100.00%), Scheduled: 682.00ms (100.00%), Output: 1096 rows (9.63kB)                                                                                                               
         Input avg.: 18262.25 rows, Input std.dev.: 173.21%                                                                                                                                               
         d_date_sk := tpcds:d_date_sk                                                                                                                                                                     
         d_year := tpcds:d_year                                                                                                                                                                           
         Input: 73049 rows (0B), Filtered: 98.50%                                                                                                                                                         
                                                                                                                                                                                                          
                                                                                                                                                                                                          
(1 row)          
```

**Comment by martint** [comment]
> Two of the TestTpcdsCostBasedPlan tests failed, due to a join order flip (log): q46 and q68.

I haven't looked at the plans, but is the new plan "better"?

**Comment by martint** [comment]
> Seems that changing the ("date_dim"."d_year" IN (1999 , (1999 + 1) , (1999 + 2))) predicate to ("date_dim"."d_year" >= 1999 AND "date_dim"."d_year" <= 1999 + 2) causes the plan change.

That seems like a bug (or inconsistency) in the cost estimator. For integer types, it's clearly equivalent.

**Comment by findepi** [comment]
> > Seems that changing the ("date_dim"."d_year" IN (1999 , (1999 + 1) , (1999 + 2))) predicate to ("date_dim"."d_year" >= 1999 AND "date_dim"."d_year" <= 1999 + 2) causes the plan change.
> 
> That seems like a bug (or inconsistency) in the cost estimator. For integer types, it's clearly equivalent.

@martint there is no logic impl'd which could realize these are equivalent.
So, while indeed equivalent, they will be hit different code paths today.

Of course, something to improve, so the obvious question would be -- how big of a problem this is, how urgent.

**Comment by sopel39** [comment]
Given that plans changes, we probably should benchmark this PR

**Comment by rzeyde-varada** [comment]
> I haven't looked at the plans, but is the new plan "better"?

I have ran the query on `sf1000` and it seems that the new plan should be more better, since the build side table is smaller than the probe side.

 - The "build side" on q46 (results in 431k rows, ~25MB):
```sql
SELECT
  "ss_ticket_number"
, "ss_customer_sk"
, "ca_city" "bought_city"
, "sum"("ss_coupon_amt") "amt"
, "sum"("ss_net_profit") "profit"
FROM
  store_sales
, date_dim
, store
, household_demographics
, customer_address
WHERE ("store_sales"."ss_sold_date_sk" = "date_dim"."d_date_sk")
   AND ("store_sales"."ss_store_sk" = "store"."s_store_sk")
   AND ("store_sales"."ss_hdemo_sk" = "household_demographics"."hd_demo_sk")
   AND ("store_sales"."ss_addr_sk" = "customer_address"."ca_address_sk")
   AND (("household_demographics"."hd_dep_count" = 4)
      OR ("household_demographics"."hd_vehicle_count" = 3))
   AND ("date_dim"."d_dow" IN (6   , 0))
   AND ("date_dim"."d_year" IN (1999   , (1999 + 1)   , (1999 + 2)))
   AND ("store"."s_city" IN ('Fairview'   , 'Midway'   , 'Fairview'   , 'Fairview'   , 'Fairview'))
GROUP BY "ss_ticket_number", "ss_customer_sk", "ss_addr_sk", "ca_city";
```

- The build side on q68 (results in 98k rows, ~7.3MB):
```
SELECT
  "ss_ticket_number"
, "ss_customer_sk"
, "ca_city" "bought_city"
, "sum"("ss_ext_sales_price") "extended_price"
, "sum"("ss_ext_list_price") "list_price"
, "sum"("ss_ext_tax") "extended_tax"
FROM
  store_sales
, date_dim
, store
, household_demographics
, customer_address
WHERE ("store_sales"."ss_sold_date_sk" = "date_dim"."d_date_sk")
   AND ("store_sales"."ss_store_sk" = "store"."s_store_sk")
   AND ("store_sales"."ss_hdemo_sk" = "household_demographics"."hd_demo_sk")
   AND ("store_sales"."ss_addr_sk" = "customer_address"."ca_address_sk")
   AND ("date_dim"."d_dom" BETWEEN 1 AND 2)
   AND (("household_demographics"."hd_dep_count" = 4)
      OR ("household_demographics"."hd_vehicle_count" = 3))
   AND ("date_dim"."d_year" IN (1999   , (1999 + 1)   , (1999 + 2)))
   AND ("store"."s_city" IN ('Midway'   , 'Fairview'));
```

- The "probe side" in both queries (results in 12M rows, ~403MB):
```sql
SELECT
  "c_last_name"
, "c_first_name"
, "ca_city"
FROM
  customer
, customer_address
WHERE ("c_current_addr_sk" = "ca_address_sk");
```

**Comment by rzeyde-varada** [comment]
Squashed and rebased over the latest `master`.

**Comment by sopel39** [comment]
@martint is it sill LGTM? I wanted to run benchmarks on this

**Comment by martint** [comment]
Yup, please do

**Comment by sopel39** [comment]
Benchmarks.
[Benchmarks merge.pdf](https://github.com/prestosql/presto/files/4758074/Benchmarks.merge.pdf)
I'm reruning them as it seems odd that merging causes CPU regression in so many queries.

However, I see two problems with this PR:
1. Filter rules based on ranger are less accurate than ones based on discrete values (e.g because for discrete values we use NDVs). This can and should be fixed for int/long types
2. Readers (e.g ORC reader: `TupleDomainOrcPredicate#extractDiscreteValues`) use discrete values to filter using Bloom filters. This should be fixed and I think this could cause visible CPU regressions

**Comment by rzeyde-varada** [comment]
Many thanks for the benchmarking!
Would you prefer to fix the problems above in separate PRs?

**Comment by findepi** [comment]
> Would you prefer to fix the problems above in separate PRs?

Would that mean we have a regression until those new PRs are done & merged?

**Comment by rzeyde-varada** [comment]
> we have a regression until those new PRs are done & merged?

No, it would be indeed undesirable.
I am suggesting first fixing those issues (in separate PRs) - and then test and make sure that this PR doesn't introduce performance regressions before merging it.

**Comment by sopel39** [comment]
Here is another benchmark.

[Benchmarks comparison-merge.pdf](https://github.com/prestosql/presto/files/4780480/Benchmarks.comparison-merge.pdf)

> I am suggesting first fixing those issues (in separate PRs) - and then test and make sure that this PR doesn't introduce performance regressions before merging it.

Good idea

**Comment by rzeyde-varada** [comment]
Marking this PR as draft, will unmark after fixing #4107 and #4108.

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Range.java
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/Range.java:231]: technically speaking, a `long`-backed type can have an ordering defined that's not compatible with ordering of `long` values used for stack representation.

it would be better to capture a relevant abstraction of a type here, eg let types define an optional method like `Optional<?> getNearestGreaterValue(Object value)`. We would implement it for intergral, decimal, date/time types. (Actually, we could implement that for real and doubles too, since effectively they are discrete types as well).

Comment (findepi) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/Range.java:231]: BTW is the current code correct for REAL type?
i.e. do

- `Long::compare`
- `(long a, long b) -> `Float.compare(Float.intBitsToFloat(toIntExact(a)), Float.intBitsToFloat(toIntExact(b)))`

define the _same_ ordering?

please add relevant tests

**Review by findepi — CHANGES_REQUESTED** [review]

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java:189]: abs should be used here, as `value < otherValue`

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java:185]: Method javadoc should be updated too

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java
Comment (findepi) [inline presto-spi/src/main/java/io/prestosql/spi/predicate/Marker.java:186]: same comment on `javaType == long.class` condition as in https://github.com/trinodb/trino/pull/3316#discussion_r681715134

File: presto-benchto-benchmarks/src/test/resources/sql/presto/tpcds/q46.plan.txt
Comment (findepi) [inline presto-benchto-benchmarks/src/test/resources/sql/presto/tpcds/q46.plan.txt:4]: Does the first commit change the plans?

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by rzeyde-varada** [comment]
Many thanks for the review!
Since the existing implementation of SortedRangeSet had changed, I will update this PR according to the comments above (hopefully in the next week).

File: presto-spi/src/main/java/io/prestosql/spi/predicate/Range.java
Comment (findepi) [reply presto-spi/src/main/java/io/prestosql/spi/predicate/Range.java:231]: see also https://github.com/trinodb/trino/issues/8893

**Comment by rzeyde-varada** [comment]
Sorry for the (very) delayed response - and many thanks for the [suggestion above](https://github.com/trinodb/trino/pull/3316#discussion_r681715134)!
I have added an `Type#areConsecutiveValues` API to allow merging a "discrete value set" for our use-case (large DFs).
This optimization is disabled by default, so it should not effect current performance (#4107 & #4108).
Please take a look and let me know your opinions :)

**Comment by findepi** [comment]
@rzeyde-varada thanks for the update

not a review yet...

> This optimization is disabled by default, so it should not effect current performance (#4107 & #4108).

I am concerned about this. Disabled optimization isn't useful for general audience. 
Who's going to use it, and under what circumstances? Would we be able to provide at least doc-level actionable guidance to when this should be enabled?

**Comment by rzeyde-varada** [comment]
> Who's going to use it, and under what circumstances? 

This optimization should be useful when the user allows [large dynamic filters](https://github.com/trinodb/trino/blob/54d8154037dfe5f6f65709dbafeb92f5506af2ac/core/trino-main/src/main/java/io/trino/execution/DynamicFilterConfig.java#L39) - where the collected DF may contain a lot of distinct values, which can be "compacted" without losing filtering selectivity (as may happen [here](https://github.com/trinodb/trino/blob/f382a176760975fec51e89f3dc62060e20420f82/core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java#L239)).

Since I was not sure if it is OK to enable it without causing other regressions (as happened [before](https://github.com/trinodb/trino/pull/3316#issuecomment-641937684)), I have added another "feature flag" to allow experimentation.

**Comment by sopel39** [comment]
What about fixing #4107 #4108 first?

**Comment by rzeyde-varada** [comment]
> What about fixing #4107 #4108 first? Would that be hard thing to do?

Not sure if it's hard, but if we enable domain compaction only for large DFs, #4107 is probably not relevant (since it was opened to fix statistics estimation).
Regarding #4108, we probably need to define some kind of "threshold" - since we probably don't want to use bloom filters for "very wide" ranges. 
For example, `[1, 1000000]` is probably not a good candidate for a bloom filter, but  `[1234, 1239]` probably is.
In addition, #4108 is less relevant since the latest version of this PR enables domain compaction only for DF (and not for any SortedRangeSet, as was done before).

**Comment by sopel39** [comment]
> In addition, #4108 is less relevant since the latest version of this PR enables domain compaction only for DF (and not for any SortedRangeSet, as was done before).

DF is just one optimization technique. Predicate pushdown (via Tuple domain) is another important one and it seems it would benefit from merging `long` values too.

> Regarding #4108, we probably need to define some kind of "threshold" - since we probably don't want to use bloom filters for "very wide" ranges.

Yes, threshold makes sense

**Comment by rzeyde-varada** [comment]
Rebased over https://github.com/trinodb/trino/pull/9868.

**Comment by findepi** [comment]
> the latest version of this PR enables domain compaction only for DF (and not for any SortedRangeSet, as was done before).

Any particular reason for this?
I would (maybe naively) assume that more compact representation in SortedRangeSet is always better.

**Comment by rzeyde-varada** [comment]
> > the latest version of this PR enables domain compaction only for DF (and not for any SortedRangeSet, as was done before).
> 
> Any particular reason for this? 

Some existing code flows handle ranges differently than discrete sets of values:
https://github.com/trinodb/trino/pull/3316#issuecomment-641937684

**Comment by rzeyde-varada** [comment]
Rebased over the latest version of https://github.com/trinodb/trino/pull/9868.

**Comment by rzeyde-varada** [comment]
Rebased over latest https://github.com/trinodb/trino/pull/9868.

**Comment by rzeyde-varada** [comment]
https://github.com/trinodb/trino/pull/9868 is merged - please take a look at this PR :)

**Review by raunaqmorarka — COMMENTED** [review]
It looks like current functionality requires multiple flags to unlock.
We will still fall back to min/max DF collection by default when there are many adjacent distinct values. So `enable-large-dynamic-filters` also has to be enabled to avoid that.
In the current implementation of DFs, connectors already receive the full Domain without simplification. We'll save some memory and network communication costs with a more compact representation in Domain. But are these efficiency gains the main benefit or is this more about preventing `Domain#simplify` call in connectors from loosing the granularity of information in the received Domain ?
Would these changes make it worthwhile to revise the default DF thresholds in DynamicFilterConfig to higher values ?

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:993]: Can we use comparison operator to detect adjacent values the way `tryMergeWithNext` does ?
That would avoid relying on `getDiscreteValues` which isn't implemented for many types.

Comment (rzeyde-varada) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:993]: Not sure - IIUC, `tryMergeWithNext()` relies on the ranges sharing the low/high bound values, i.e. `[1,5]` and `(5,9]` can be merged into `[1,9]`.
However, if we try to merge `[1,5]` and `[6,9]`, we should make sure that 5 and 6 are "consecutive" (i.e. there is no other value between them, and this requires calling `Type#getDiscreteValues`).

**Comment by rzeyde-varada** [comment]
Sorry for the delayed response - fixed the issues, and updated the PR.

> But are these efficiency gains the main benefit or is this more about preventing `Domain#simplify` call in connectors from loosing the granularity of information in the received Domain ?

We would prefer to not to call `Domain#simplify` to allow applying the DF efficiently.
This way, our connector will get an "equivalent" predicate, which can be evaluated much faster (compared to a large discrete values' list).

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:195]: Why are the values being multiplied by 10 in this test and in testCollectWithNulls ?

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:976]: If we initialised `current` to the first value in `ranges.getOrderedRanges()`, could we avoid using `Optional` here ?

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java
Comment (raunaqmorarka) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:998]: Yes, we can't use similar logic as `tryMergeWithNext` here.
But re-using `Type#getDiscreteValues` also seems awkward.
I'm thinking we should just add a `boolean areConsecutive(Object low, Object high)` to Type for this.
@findepi @martint thoughts ?

File: core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: Document whether the compacted version is supported to 

- be equal (as in `Object.equals`)
- contain same information, or can eg be slightly lossy (widening) for more efficient compaction
- what does is returned when compaction is not possible?

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: Also, higher level

- why is this a method on `Ranges` instead of on `ValueSet`?
- why do we want to have such method at all? I would hope the `ValueSet` instance is "as compact as possible", always

**Review by findepi — CHANGES_REQUESTED** [review]

File: core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java
Comment (rzeyde-varada) [reply core/trino-main/src/test/java/io/trino/operator/TestDynamicFilterSourceOperator.java:195]: To prevent them from being merged.
If you prefer, I can keep the original values - and change the expected results to check that discrete values' merging works.

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java
Comment (rzeyde-varada) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:976]: Sounds good, will fix.

File: core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java
Comment (rzeyde-varada) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: > Document whether the compacted version is supported to

Sounds good - done.

Comment (rzeyde-varada) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: Thinking about it, the relevant part of this PR (https://github.com/trinodb/trino/pull/3316/commits/23d6a6e91be03de75d0c5ed152447439e4de991b#diff-a3ede16a05943d869ba6d7b8821761bc3ed3d23915bb0b3bf8bac363b7038dea) is small enough to just move into our connector (after https://github.com/trinodb/trino/pull/9868 is merged).

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: > I agree, but such compactions [may cause CBO to change the plan](https://github.com/trinodb/trino/pull/3316#issuecomment-608980527) (since we estimate the statistics differently depending on the predicate being a range or a discrete set of values)

Is this something that could be improved on?

cc @sopel39

Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/Ranges.java:35]: > Is this something that could be improved on?

IMO it should be, I don't think there should be code duality (DF vs CBO)

File: core/trino-spi/src/main/java/io/trino/spi/predicate/ValueSet.java
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/ValueSet.java]: > The concept of compacting ranges seems very specific to SortedRangeSet

agreed

note however, that the interface
```
/**
 * Try to return a more compact representation (if possible).
 */
ValueSet compact();
```

is not range-specific and could live in ValueSet.
Current name is `getCompactedRanges`, but it still is `ValueSet`-breaking operation, which _could_ be potentially extended to other ValueSet types.

I think the important question to ask is what "compact" actually means?

- occupies less memory? (eg SortedRangeSet uses dictionary for singleton-ranges, but only if constructed explicitly as such)
   - in the first case, it's not ranges-specific
- fewer ranges?
  - in the second case, it's ranges specific, but then why does it return a ValueSet, losing information it's ranges?

File: core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:995]: While working on Iceberg I had a need to find an adjacent value (not check whether two values are adjacent).
Here is my API proposal for that: https://github.com/trinodb/trino/pull/12797

Here, it seems we don't need new methods on Type, so I'd hold off and not add them.

**Comment by colebow** [comment]
:wave: @rzeyde-varada - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Pull request closed without merging**
